pep-0592.rst

```diff
@@ -0,0 +1,280 @@
+cases for annotations described in PEP 3107 (database mapping,
+foreign languages bridge) are not currently realistic given the
+prevalence of type annotations.
+
+Rationale
+---------
+
+We propose adding an ``Annotated`` type to the typing module to decorate
+existing types with context-specific metadata. Specifically, a type
+``T`` can be annotated with metadata ``x`` via the typehint
+``Annotated[T, x]``. This metadata can be used for either static
+analysis or at runtime. If a library (or tool) encounters a typehint
+``Annotated[T, x]`` and has no special logic for metadata ``x``, it
+should ignore it and simply treat the type as ``T``. Unlike the
+``no_type_check`` functionality that current exists in the ``typing``
+module which completely disables typechecking annotations on a function
+or a class, the ``Annotated`` type allows for both static typechecking
+of ``T`` (e.g., via MyPy or Pyre, which can safely ignore ``x``)
+together with runtime access to ``x`` within a specific application. We
+believe that the introduction of this type would address a diverse set
+of use cases of interest to the broader Python community.
+
+This was originally brought up as issue 600 [issue-600]_ in the typing github
+and then discussed in Python ideas [python-ideas]_.
+
+Motivating examples
+~~~~~~~~~~~~~~~~~~~
+
+Reading binary data
+^^^^^^^^^^^^^^^^^^^
+
+The ``struct`` module provides a way to read and write C structs
+directly from their byte representation. It currently relies on a string
+representation of the C type to read in values::
+
+    record = b'raymond   \x32\x12\x08\x01\x08'
+    name, serialnum, school, gradelevel = unpack('<10sHHb', record)
+
+The struct documentation [struct-doc]_ suggests using a named tuple to unpack
+the values and make this a bit more tractable::
+
+    from collections import namedtuple
+    Student = namedtuple('Student', 'name serialnum school gradelevel')
+    Student._make(unpack('<10sHHb', record))
+    # Student(name=b'raymond   ', serialnum=4658, school=264, gradelevel=8)
+
+However, this recommendation is somewhat problematic; as we add more
+fields, it’s going to get increasingly tedious to match the properties
+in the named tuple with the arguments in ``unpack``.
+
+Instead, annotations can provide better interoperability with a type
+checker or an IDE without adding any special logic outside of the
+``struct`` module:
+
+::
+
+    from typing import NamedTuple
+    UnsignedShort = Annotated[int, struct.ctype('H')]
+    SignedChar = Annotated[int, struct.ctype('b')]
+
+    @struct.packed
+    class Student(NamedTuple):
+      # MyPy typechecks 'name' field as 'str' 
+      name: Annotated[str, struct.ctype("<10s")]
+      serialnum: UnsignedShort
+      school: SignedChar
+      gradelevel: SignedChar
+
+    # 'unpack' only uses the metadata within the type annotations
+    Student.unpack(record)
+    # Student(name=b'raymond   ', serialnum=4658, school=264, gradelevel=8)
```

Comment:
This example (and to certain extent the second example) is unnecessarily detailed. You focus too much on advocating for specific use case, instead of explaining the big picture with the example.